sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -156,7 +171,15 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
       // If a CTE relation is never used, it will disappear after inline. Here we explicitly check
       // analysis for it, to make sure the entire query plan is valid.
       try {
-        if (refCount == 0) checkAnalysis0(relation.child)
```

Comment:
Do you think we can keep this check as it is, but call the already existing `cleanCTEMap()` right after `buildCTEMap()` a bit above? `cleanCTEMap()`'s comment seems wrong and it doesn't remove CTE defs without useful refs, but it just set ref counts to 0.

Reply:
I think the current code is better as it guarantees the leaf CTE relations (do not reference any other) are checked first. Maybe we can make it clearer using comments

Reply:
Ok I added some comments to explain we are trying to check the leaf relation first because that gives the most accurate error message.

Reply:
Sorry, for the late reply. (I was travelling this week.)

The reason why I would suggest the `cleanCTEMap()` approach is because you might be checking a CTE 2 times now. Not in this new code, as `visited` map makes sure to check a CTE only once, but later when we check the inlined plan it might contain a CTE that was referenced from an other CTE with ref count = 0 (so the CTE is checked in the new code) and also from a third CTE with valid references (so the CTE is checked in the inlined plan again).

As for checking the leafs first, checking the CTEs in id order guarantees that. This is because how we do CTE resolution (assign ids to defs) in `CTESubstitution`. `cleanCTEMap()` also uses this property to quickly clear useless transitive reference counts.